pkg/storage/stores/indexshipper/compactor/retention/retention.go

```diff
@@ -90,7 +227,17 @@ type Marker struct {
 	markTimeout      time.Duration
 }
 
+var mu sync.Mutex
```

Comment:
I don't think we need the mutex here. The compactor should probably own the marker metrics and pass it to markers.

Reply:
I did not implement this.
This section of the code was written by @rbarry82 who sadly passed away a month ago.

Do you mean something like this?

```golang
func NewMarker(workingDirectory string, expiration ExpirationChecker, markTimeout time.Duration, chunkClient client.Client, metrics *markerMetrics) (*Marker, error) {
	return &Marker{
		workingDirectory: workingDirectory,
		expiration:       expiration,
		markerMetrics:    metrics,
		chunkClient:      chunkClient,
		markTimeout:      markTimeout,
	}, nil
}
```

Reply:
I'm sorry to hear about the loss of your colleague :slightly_frowning_face: 

Metrics can only be registered once. If we try to register them a second time, a runtime panic occurs. The code, as is prevents registering a second time but it's a little akward.

I think we could change
```go
func NewMarker(workingDirectory string, expiration ExpirationChecker, markTimeout time.Duration, chunkClient client.Client, r prometheus.Registerer) (*Marker, error)
```
to take metrics like this
```
func NewMarker(workingDirectory string, expiration ExpirationChecker, markTimeout time.Duration, chunkClient client.Client, metrics *markerMetrics) (*Marker, error)
```

But then we need a place to instantiate the metrics. We could probably do it in the compactor's `init` function and make it a field on the compactor.

If we do this, the mutex can be deleted altogether.

Reply:
About mutex, now I remember that Ryan added [the following to `Compactor` struct:](https://github.com/grafana/loki/blob/69f78bee44ba8d2f5a223243a2fdd2b100cad692/pkg/storage/stores/indexshipper/compactor/compactor.go#L182)

```golang
	// Size based compaction means that two compactions might try to happen at the same time.
	// Use this to ensure size-based and normal compaction can't step on eachother.
	compactionMtx sync.Mutex
```

Does make sense o he was wrong?



About your suggestion, I have this draft right now (didn't push it yet)... Is it close to what you suggest?


```diff
modified   pkg/storage/stores/indexshipper/compactor/compactor.go
@@ -182,6 +182,7 @@ type Compactor struct {
 	// Size based compaction means that two compactions might try to happen at the same time.
 	// Use this to ensure size-based and normal compaction can't step on eachother.
 	compactionMtx sync.Mutex
+	markerMetrics *retention.markerMetrics
 
 	// one for each object store
 	storeContainers map[string]storeContainer
@@ -310,6 +311,8 @@ func (c *Compactor) init(objectStoreClients map[string]client.ObjectClient, sche
 		}
 	}
 
+	c.markerMetrics = retention.newMarkerMetrics(r)
+
 	c.storeContainers = make(map[string]storeContainer, len(objectStoreClients))
 	for objectStoreType, objectClient := range objectStoreClients {
 		var sc storeContainer
@@ -338,7 +341,7 @@ func (c *Compactor) init(objectStoreClients map[string]client.ObjectClient, sche
 				return fmt.Errorf("failed to init sweeper: %w", err)
 			}
 
-			sc.tableMarker, err = retention.NewMarker(retentionWorkDir, c.expirationChecker, c.cfg.RetentionTableTimeout, chunkClient, r)
+			sc.tableMarker, err = retention.NewMarker(retentionWorkDir, c.expirationChecker, c.cfg.RetentionTableTimeout, chunkClient, c.markerMetrics)
 			if err != nil {
 				return fmt.Errorf("failed to init table marker: %w", err)
 			}
modified   pkg/storage/stores/indexshipper/compactor/retention/retention.go
@@ -5,9 +5,9 @@ import (
 	"context"
 	"errors"
 	"fmt"
-	"sync"
 	"os"
 	"path/filepath"
+	"sync"
 
 	"time"
 
@@ -227,21 +227,11 @@ type Marker struct {
 	markTimeout      time.Duration
 }
 
-var mu sync.Mutex
-var metrics *markerMetrics
-
-func NewMarker(workingDirectory string, expiration ExpirationChecker, markTimeout time.Duration, chunkClient client.Client, r prometheus.Registerer) (*Marker, error) {
-	mu.Lock()
-	defer mu.Unlock()
-
-	if metrics == nil {
-		metrics = newMarkerMetrics(r)
-	}
-
+func NewMarker(workingDirectory string, expiration ExpirationChecker, markTimeout time.Duration, chunkClient client.Client, metrics *markerMetrics) (*Marker, error) {
 	return &Marker{
 		workingDirectory: workingDirectory,
 		expiration:       expiration,
-		markerMetrics:    newMarkerMetrics(r),
+		markerMetrics:    metrics,
 		chunkClient:      chunkClient,
 		markTimeout:      markTimeout,
 	}, nil
```